### ISSUE TITLE
Compatibility with Tensors

### ISSUE THREAD

**Issue by pietercolpaert** [KNowledgeOnWebScale/rdf-timeseriessnippets#12]
See this spec: https://neverblink-oss.github.io/rdf-tensor/

This spec has a very similar idea to the idea that we had over here. However, it also introduces SPARQL functions specifically for doing tensor operations where GPU operations could be leveraged: https://onnx.ai/onnx/operators/

I propose to open up a line of work to:
 1. Study how we could use a [NumericDataTensor](https://w3id.org/rdf-tensor/vocab#NumericDataTensor) and this way make this spec entirely compatible with RDF tensors
 2. Study what functions we need on top of the time series, and whether these can be expressed as a combination of ONNX operators (which RDF Tensors will be in line with)

Where it will shine is when also using a binary compression, that could be published in a ttl document using base64 encoding

**Comment by Ostrzyciel** [comment]
Writing down here the full context of what we discussed in Ghent with Pieter:

- RDF-tensor will very soon switch to using ONNX IR for defining the behavior of its SPARQL operators (lead dev: @nk2ishere).
  - ONNX IR is a portable intermediate representation for expressing linear algebra and ML operations. It's essentially an industry standard, stewarded by the Linux Foundation.
  - ONNX IR can be leveraged by runtimes available for every major programming language, including JavaScript (in the browser): https://onnxruntime.ai/docs/tutorials/web/
  - ONNX IR is compiled to a number of native instruction sets by the runtime. You can use translate it to CUDA, x86 AVX, ARM Neon, various NPUs and more: https://onnxruntime.ai/docs/execution-providers/
  - In that way, ONNX provides a vendor-agnostic way of expressing numerical computation graphs that can be compiled to any platform on runtime.
  - ONNX IR can be written manually or translated automatically from a high-level framework like Keras or TensorFlow.
  - We chose this approach, because:
    - It delegates testing/compatibility issues to the ONNX spec – we don't need extensive test suites.
    - Implementing our SPARQL functions is as easy as downloading the ONNX IR fragments and passing them to the relevant ONNX runtime. This also makes any spec changes or updates trivial to handle by implementations.
    - We get full native performance and hardware acceleration.

- Research question 1: can time series be expressed as a NumericDataTensor? What we've found so far:
  - This would have the advantage of facilitating very efficient exchange of dense time series data and enable us to use accelerated processing.
  - For multi-column time series, they can be merged into a single tensor, as long as all columns use the same datatype.
  - For heterogeneous datasets (different columns have different datatypes), each column must be a separate tensor.
  - Timestamps can be encoded as int64, however additional RDF metadata would be required to explain what is the unit (seconds, millis, micros, nanos) and what is the zero point (e.g., 1970-01-01 for Unix timestamps).
  - Representing non-numerical and non-timestamp data is not possible in tensors. However:
    - Pieter said that we don't really have a strong use case for storing IRIs within time series, this was more of an academic experiment. I agree.
    - Theoretically we could use CDTs for this purpose, although they are much more open in scope (e.g., they allow blank nodes) which can create extra headaches.
    - In summary, it's probably best to discard this feature.
  - This kind of columnar storage is very well aligned with how modern TSDBs (time series databases) work, as they most often use columnar layouts and extensive column-wise compression.
  - The value space of these literals is a pure numerical tensor, uncompressed, with no delta encoding. Delta encoding, compression, or whatever else can be applied in **lexical space** of the tensor literal.
    - In that sense, the tensors would have many possible lexical representations, depending on the use case / available protocols.
    - This creates a fragmentation issue, though. We will have to look into what common (preferably standardized) lexical representations of numerical arrays could be mandated as "preffered" or "standard" to foster interoperability.
    - Binary representations can be embedded in text-based RDF formats using base64 encoding. This should be much more efficient than JSON strings.

- Research question 2: which timeseries functions do we need? How to implement them on top of RDF tensors?
  - As long as the relevant function can be expressed in ONNX IR, we can reuse the machinery set up in the RDF tensor spec.
  - We can simply say that "this is an extension of the RDF tensor spec, we just define some new SPARQL functions using ONNX IR". Then implementations of RDF Tensor can simply take these ONNX IR fragments, register them and boom, you have time series support.
  - We have identified some potential function collections that may be useful:
    - https://time-series-features.gitbook.io/catch22
    - https://stumpy.readthedocs.io/en/latest/index.html
  - We don't know yet if they can be implemented in ONNX – this is to be investigated. If not, we may resort to a hybrid algorithm (e.g., pseudocode with tensor variables and operations expressed in ONNX IR).
